### PR TITLE
Add ability to WPT-test file uploads and fetches, fixes #12322

### LINKED ISSUE #12322: Add ability to WPT-test file uploads and fetches
We should perhaps have a command line option or environment variable that specifies a file to upload, and use that for writing Servo-specific WPT tests that upload files and test that they get fetched properly.

### PR DESCRIPTION
**Pull request by izgzhen**
Using `inputElem.selectFiles(["path1", "path2"])` in JavaScript with pref `dom.htmlinputelement.select_files.enabled` = `true` will simulate the effect of `inputElem.click()`.

See #12322  for more, also related to #12343, #11131.

---

This change is [Reviewable](https://reviewable.io/reviews/servo/servo/12344)

### CODE REVIEW

File: components/script/dom/webidls/HTMLInputElement.webidl [hunk 1]
```diff
@@ -70,6 +70,11 @@ interface HTMLInputElement : HTMLElement {
   void setSelectionRange(unsigned long start, unsigned long end, optional DOMString direction);
 
   // also has obsolete members
+
+  // Select with file-system paths for testing purpose
+  [Pref="dom.htmlinputelement.select_files.enabled"]
```
Comment (Manishearth) [inline components/script/dom/webidls/HTMLInputElement.webidl]: Perhaps this should be under dom.testing?

File: components/script/dom/htmlinputelement.rs [hunk 2]
```diff
@@ -731,6 +741,80 @@ impl HTMLInputElement {
         let el = self.upcast::<Element>();
         el.set_placeholder_shown_state(has_placeholder && !has_value);
     }
+
+    // https://html.spec.whatwg.org/multipage/#file-upload-state-(type=file)
+    // Select files by invoking UI or by passed in argument
+    fn select_files(&self, opt_test_paths: Option<Vec<DOMString>>) {
```
Comment (Manishearth) [inline components/script/dom/htmlinputelement.rs]: This isn't process safe -- we should assert that the pref is enabled if opt_test_paths is nonempty here.

File: components/net/filemanager_thread.rs [hunk 3]
```diff
@@ -231,8 +238,13 @@ impl<UI: 'static + UIProvider> FileManager<UI> {
 
     fn select_files(&mut self, patterns: Vec<FilterPattern>,
                     sender: IpcSender<FileManagerResult<Vec<SelectedFile>>>,
-                    origin: FileOrigin) {
-        match self.ui.open_file_dialog_multi("", patterns) {
+                    origin: FileOrigin, opt_test_paths: Option<Vec<String>>) {
+        let opt_v = match opt_test_paths {
+            Some(v) => Some(v),
+            None => self.ui.open_file_dialog_multi("", patterns),
+        };
```
Comment (Manishearth) [inline components/net/filemanager_thread.rs]: This isn't process safe -- we should assert that the pref is enabled if opt_test_paths is nonempty here.

**Comment by izgzhen** [comment]
Okay, this passed the CI as well

File: components/net/filemanager_thread.rs [hunk 4]
```diff
@@ -212,8 +215,15 @@ impl<UI: 'static + UIProvider> FileManager<UI> {
 
     fn select_file(&mut self, patterns: Vec<FilterPattern>,
                    sender: IpcSender<FileManagerResult<SelectedFile>>,
-                   origin: FileOrigin) {
-        match self.ui.open_file_dialog("", patterns) {
+                   origin: FileOrigin, opt_test_path: Option<String>) {
+        let select_files_enabled = PREFS.get("dom.testing.htmlinputelement.select_files.enabled").as_boolean();
+
+        let opt_s = match (opt_test_path, select_files_enabled) {
+            (Some(s), Some(true)) => Some(s),
```
Comment (Manishearth) [inline components/net/filemanager_thread.rs]: we can just match on `Some(s) if select_files_enabled` here

Comment (izgzhen) [inline components/net/filemanager_thread.rs]: Well but `select_files_enabled` is `Option<boo>` :/

Comment (Manishearth) [inline components/net/filemanager_thread.rs]: .as_boolean().unwrap_or(false)

File: components/net/filemanager_thread.rs [hunk 5]
```diff
@@ -309,8 +310,14 @@ impl <UI: 'static + UIProvider> FileManagerStore<UI> {
...
     fn select_file(&self, patterns: Vec<FilterPattern>,
                    sender: IpcSender<FileManagerResult<SelectedFile>>,
-                   origin: FileOrigin) {
-        match self.ui.open_file_dialog("", patterns) {
+                   origin: FileOrigin, opt_test_path: Option<String>) {
+        let opt_s = if select_files_pref_enabled() {
+            opt_test_path
+        } else {
+            self.ui.open_file_dialog("", patterns)
+        };
```
Comment (izgzhen) [inline components/net/filemanager_thread.rs]: @Manishearth Looks good?

Comment (Manishearth) [inline components/net/filemanager_thread.rs]: Yep. Leave a comment here though mentioning that while this pref check is redundant, it's in place to ensure that a compromised script thread cannot request arbitrary files. r=me

File: components/net/filemanager_thread.rs [hunk 6]
```diff
@@ -328,8 +337,16 @@ impl <UI: 'static + UIProvider> FileManagerStore<UI> {
 
     fn select_files(&self, patterns: Vec<FilterPattern>,
                     sender: IpcSender<FileManagerResult<Vec<SelectedFile>>>,
-                    origin: FileOrigin) {
-        match self.ui.open_file_dialog_multi("", patterns) {
+                    origin: FileOrigin, opt_test_paths: Option<Vec<String>>) {
+        // Check if the select_files preference is enabled,
+        // if so applying opt_test_paths for testing convenience
```
Comment (izgzhen) [inline components/net/filemanager_thread.rs]: @Manishearth done

Comment (izgzhen) [inline components/net/filemanager_thread.rs]: @Manishearth wait ... I will add another line about security

**Pull request merged**